src/EditorFeatures/CSharpTest/ReassignedVariable/CSharpReassignedVariableTests.cs

```diff
@@ -0,0 +1,853 @@
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    int this[int [|p|]] { set => [|p|]++; }
+}");
+        }
+
+        [Fact]
+        public async Task TestIndexerWithWriteInSetter2()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    int this[int [|p|]] { set { [|p|]++; } }
+}");
+        }
+
+        [Fact]
+        public async Task TestPropertyWithAssignmentToValue1()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    int Goo { set => [|value|] = [|value|] + 1; }
+}");
+        }
+
+        [Fact]
+        public async Task TestPropertyWithAssignmentToValue2()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    int Goo { set { [|value|] = [|value|] + 1; } }
+}");
+        }
+
+        [Fact]
+        public async Task TestEventAddWithAssignmentToValue()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    event Action Goo { add { [|value|] = null; } remove { } }
+}");
+        }
+
+        [Fact]
+        public async Task TestEventRemoveWithAssignmentToValue()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    event Action Goo { add { } remove { [|value|] = null; } }
+}");
+        }
+
+        [Fact]
+        public async Task TestLambdaParameterWithoutReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        Action<int> a = x => Console.WriteLine(x);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestLambdaParameterWithReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        Action<int> a = [|x|] => Console.WriteLine([|x|]++);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestLambdaParameterWithReassignment2()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        Action<int> a = (int [|x|]) => Console.WriteLine([|x|]++);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestLocalWithoutInitializerWithoutReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(bool b)
+    {
+        int p;
+        if (b)
+            p = 1;
+        else
+            p = 2;
+
+        Console.WriteLine(p);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestLocalWithoutInitializerWithReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(bool b)
+    {
+        int [|p|];
+        if (b)
+            [|p|] = 1;
+        else
+            [|p|] = 2;
+
+        [|p|] = 0;
+        Console.WriteLine([|p|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestLocalDeclaredByPattern()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        if (0 is var [|p|]) [|p|] = 0;
```

Comment:
❔ Do we already have tests for the other path (where `p` is written in the `else` instead of the `if`)?

```
void M(int? x)
{
  if (x is int p)
  {
  }
  else
  {
    p = 0;
  }
}
```

Reply:
added.